felix/fv/rpf_test.go

```diff
@@ -0,0 +1,161 @@
+				log.WithField("policy", dumpResource(policy)).Info("Creating policy")
+				policy, err := calicoClient.GlobalNetworkPolicies().Create(utils.Ctx, policy, utils.NoOptions)
+				Expect(err).NotTo(HaveOccurred())
+				return policy
+			}(pol)
+
+			cc = &Checker{
+				CheckSNAT: true,
+				Protocol:  "udp",
+			}
+		})
+
+		JustAfterEach(func() {
+			if CurrentGinkgoTestDescription().Failed {
+				for _, felix := range felixes {
+					felix.Exec("iptables-save", "-c")
+					felix.Exec("ip", "link")
+					felix.Exec("ip", "addr")
+					felix.Exec("ip", "rule")
+					felix.Exec("ip", "route")
+				}
+			}
+		})
+
+		It("should not allow packets from wrong direction with non-strict RPF on main device", func() {
+			By("turning off RPF per device", func() {
+				felixes[0].Exec("sysctl", "-w", "net.ipv4.conf.all.rp_filter=0")
+				felixes[0].Exec("sysctl", "-w", "net.ipv4.conf.default.rp_filter=0")
+				felixes[0].Exec("sysctl", "-w", "net.ipv4.conf.eth0.rp_filter=0")
+			})
+
+			var eth20 *workload.Workload
+
+			By("setting up node's fake external ifaces", func() {
+				// We name the ifaces ethXY since such ifaces are
+				// treated by felix as external to the node
+				//
+				// Using a test-workload creates the namespaces and the
+				// interfaces to emulate the host NICs
+
+				eth20 = &workload.Workload{
+					Name:          "eth20",
+					C:             felixes[0].Container,
+					IP:            "192.168.20.1",
+					Ports:         "57005", // 0xdead
+					Protocol:      "udp",
+					InterfaceName: "eth20",
+				}
+				err := eth20.Start()
+				Expect(err).NotTo(HaveOccurred())
+
+				// assign address to eth20 and add route to the .20 network
+				felixes[0].Exec("ip", "route", "add", "192.168.20.0/24", "dev", "eth20")
+				felixes[0].Exec("ip", "addr", "add", "10.0.0.20/32", "dev", "eth20")
+				_, err = eth20.RunCmd("ip", "route", "add", "10.0.0.20/32", "dev", "eth0")
+				Expect(err).NotTo(HaveOccurred())
+				// Add a route to the test workload to the fake external
+				// client emulated by the test-workload
+				_, err = eth20.RunCmd("ip", "route", "add", w.IP+"/32", "via", "10.0.0.20")
+				Expect(err).NotTo(HaveOccurred())
+
+				// Make sure that networking with the .20 network works
+				cc.ResetExpectations()
+				cc.Expect(Some, eth20, w)
+				cc.CheckConnectivity()
+			})
+
+			By("testing with bad source", func() {
+				fakeWorkloadIP := "10.65.15.15"
+
+				tcpdump := w.AttachTCPDump()
+				tcpdump.SetLogEnabled(true)
+				matcher := fmt.Sprintf("IP %s\\.30446 > %s\\.30446: UDP", fakeWorkloadIP, w.IP)
+				tcpdump.AddMatcher("UDP-30446", regexp.MustCompile(matcher))
+				tcpdump.Start()
+				defer tcpdump.Stop()
+
+				_, err := eth20.RunCmd("/pktgen", fakeWorkloadIP, w.IP, "udp",
+					"--port-src", "30446", "--port-dst", "30446", "--ip-id", "666")
+				Expect(err).NotTo(HaveOccurred())
+
+				// Expect to receive the packet from the .20 as the routing is correct
```

Comment:
Comment doesn't seem to match what the check is doing.  (I.e. that there isn't anything in tcpdump)